distributed/scheduler.py

```diff
@@ -7708,6 +7674,115 @@ def _validate_ready(state: SchedulerState, ts: TaskState) -> None:
     assert all(dts.who_has for dts in ts.dependencies)
 
 
+def _queueable_to_processing(
+    state: SchedulerState, ts: TaskState
+) -> tuple[Recs, dict[str, list[Any]], dict[str, list[Any]]]:
+    "Common logic for transitioning a queueable (root-ish) task to processing, along with the rest of its family"
+    # Fastpath: skip family search and everything else if no workers are free.
+    # Since all siblings are assigned at once, this means `ts` belongs to a family we haven't processed yet.
+    # This means that we've already filled the cluster with root task families, and shouldn't schedule any more.
+    if not state.idle:
+        return {ts.key: "queued"}, {}, {}
+
+    # TODO maxsize as config/what?!
+    fam = family(ts, maxsize=20, widely_shared_cutoff=len(state.workers))
```

Comment:
I don't get why `widely_shared_cutoff=len(state.workers)`. At least when reading the code of `family` I get the impression that this parameter is used in many different places, partially to restrict runtime complexity, partially to assess parallelism. I don't think a "family" should be mixed up with cluster size.

I do believe we can and should differentiate between a static graph analysis and a dynamic assessment of potential of parallelism.

A widely shared dependency, e.g. one that is used by all root-ish tasks is a widely shared dependency no matter how large the cluster is.

If we were able to remove this dynamic component we could calculate this family once at update_graph time and be done with it, reducing the concerns about runtime complexity significantly.

Reply:
> we could calculate this family once at update_graph time and be done with it, reducing the concerns about runtime complexity significantly

It would probably be good to be able to cache the family and never have to recompute it. However, it may not be good to put it all in `update_graph`. `update_graph`, and things like `dask.order` within it, already can block the event loop for extremely long times (easily 30s with large graphs).

One thing I like about the approach here is how it spreads out the family traversal over time, and defers it until it's actually necessary.

Reply:
>  "a dependency whose current location should be ignored during scheduling, because in the process of scheduling all its dependents (without leaving workers idle), it will end up being copied to every worker anyway."

I accept that this is not straight forward and I honestly don't care about what magic number we use for initial prototyping. I care mostly about this number being static vs dynamic.

For instance, what you are describing is something like a "max-breadth" metric which I'm pretty sure we can get out of something in dask.order. Counting the number of root tasks (not root-ish, i.e. zero dependencies) would also be trivial. No matter what we're using here, this is static. I'm just trying to understand what is a static and what is a dynamic parameter and it would be helpful for me if we can use different variables for the cases where it's static vs dynamic. At the very least a comment in the code.

> I like this idea overall though. We may choose what to do with a family dynamically, based on cluster size. But I hope we can make the definition of "what is a family" static. I will look into that.

Exactly!
Random thoughts without any deeper reflection
- calculate the total number of detected families and compare with number of workers
- size of siblings vs workers

> It would probably be good to be able to cache the family and never have to recompute it. However, it may not be good to put it all in update_graph. update_graph, and things like dask.order within it, already can block the event loop for extremely long times (easily 30s with large graphs).

Sure, maybe. If we agree this to be a static component we can cache it. If we cache this right away by computing it in update_graph and assigning it to the task or by calculating it on demand is mostly an implementation detail.
The one problem I encounter here either way is cache invalidation. Whenever a new update_graph is called we'd likely need to recompute the entire graph (at least if the two graphs are connected)

> One thing I like about the approach here is how it spreads out the family traversal over time, and defers it until it's actually necessary.

Yes, I get that. There are two problems we're talking about right now
1. Finding a good algorithm for the decision. In this instance, I care about the static component of this calculation the most.
2. Making sure that the execution of this algorithm is fast and smooth (runtime complexity, event loop blocking, etc.)

A counter argument for JIT computation (in decide*/transition) is that this increases latencies for scheduling decisions compared to AOT (in update_graph). I could see both being valuable. Moving it from the one to the other should be easy enough which is why I care much more about 1.)